sys/include/net/sock/dtls.h

```diff
@@ -0,0 +1,556 @@
+ *
+ *     if (!ipv6_addr_from_str((ipv6_addr_t *)remote.addr.ipv6, addr_str)) {
+ *         puts("Error parsing destination address");
+ *         return;
+ *     }
+ *
+ *     res = sock_dtls_create(&dtls_sock, &udp_sock, DTLS_SOCK_CLIENT_TAG, 0);
+ *     if (res < 0) {
+ *         puts("Error creating DTLS sock");
+ *         return;
+ *     }
+ *
+ *     res = sock_dtls_establish_session(&dtls_sock, &remote, &session);
+ *     if (res < 0) {
+ *         printf("Error establishing session: %d\n", (int)res);
+ *         goto end;
+ *     }
+ *
+ *     res = sock_dtls_send(&dtls_sock, &session, data, datalen);
+ *     if (res < 0) {
+ *         printf("Error sending DTLS message: %d\n", (int)res);
+ *         goto end;
+ *     }
+ *
+ *     res = sock_dtls_recv(&dtls_sock, &session, rcv, sizeof(rcv), SOCK_NO_TIMEOUT);
+ *     if (res < 0) {
+ *         printf("Error receiving DTLS message: %d\n", (int)res);
+ *         goto end;
+ *     }
+ *
+ *     end:
+ *         puts("Terminating");
+ *         sock_dtls_terminate_session(&dtls_sock, &session);
+ *         sock_dtls_close(&dtls_sock);
+ *  }
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ * This is an example of a DTLS echo client.
+ *
+ * Like the server, we must first create the UDP sock.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * sock_udp_t udp_sock;
+ * sock_udp_ep_t local = SOCK_IPV6_EP_ANY;
+ * local.port = 12345;
+ * sock_udp_create(&udp_sock, &local_ep, NULL, 0);
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After that, we set the address of the remote endpoint where the packet is
+ * supposed to be sent to and the port the server is listening to which is
+ * DTLS_DEFAULT_PORT(20220).
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * sock_udp_ep_t remote;
+ * remote.port = DTLS_DEFAULT_PORT;
+ * remote.netif = gnrc_netif_iter(NULL)->pid;   // only if gnrc_netif_numoff == 1
+ * if (!ipv6_addr_from_str((ipv6_addr_t *)remote.addr.ipv6, addr_str)) {
+ *     puts("Error parsing destination address");
+ *     return;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * After the UDP sock is created, we can proceed with creating the DTLS sock.
+ * Before sending the packet, we must first establish a session with the remote
+ * endpoint using @ref sock_dtls_establish_session(). If the handshake is
+ * successfull and the session is established, we can use its `remote` to send
+ * packets to it with @ref sock_dtls_send().
+ *
+ * If error during any of these operation, the session is terminated and the sock
+ * is closed using @ref sock_dtls_terminate_session() and
+ * @ref sock_dtls_close().
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * res = sock_dtls_create(&dtls_sock, &udp_sock, DTLS_SOCK_CLIENT_TAG, 0);
+ * if (res < 0) {
+ *     puts("Error creating DTLS sock");
+ *     return;
+ * }
+ *
+ * res = sock_dtls_establish_session(&dtls_sock, &remote, &session);
+ * if (res < 0) {
+ *     printf("Error establishing session: %d\n", (int)res);
+ *     goto end;
```

Comment:
Please don't advertise `goto`s in the high-level doc.

Reply:
I'll rewrite this without using `goto`s.